tests/garage/tf/regressors/test_categorical_mlp_regressor_with_model.py

```diff
@@ -0,0 +1,244 @@
+import pickle
+from unittest import mock
+
+from nose2.tools.params import params
+import numpy as np
+import tensorflow as tf
+
+from garage.tf.optimizers import ConjugateGradientOptimizer, LbfgsOptimizer
+from garage.tf.regressors import CategoricalMLPRegressorWithModel
+from tests.fixtures import TfGraphTestCase
+
+
+def get_labels(input_shape, xs):
+    label = [0, 0]
+    if input_shape == (1, ):
+        ys = 0 if np.sin(xs) <= 0 else 1
+        label[ys] = 1
+
+    elif input_shape == (2, ):
+        ys = int(np.round(xs[0])) ^ int(np.round(xs[1]))
+        label[ys] = 1
+
+    return label
+
+
+def get_train_data(input_shape):
+    if input_shape == (1, ):
+        data = np.linspace(-np.pi, np.pi, 1000)
+        obs = [{
+            'observations': [[x]],
+            'returns': [get_labels(input_shape, x)]
+        } for x in data]
+
+    elif input_shape == (2, ):
+        data = [np.random.rand(2) for _ in range(1000)]
+        obs = [{
+            'observations': [x],
+            'returns': [get_labels(input_shape, x)]
+        } for x in data]
+    return obs
+
+
+def get_test_data(input_shape):
+    if input_shape == (1, ):
+        paths = {
+            'observations': [[-np.pi / 2], [-np.pi / 3], [-np.pi / 4],
+                             [np.pi / 4], [np.pi / 3], [np.pi / 4]]
+        }
+        expected = [[1, 0], [1, 0], [1, 0], [0, 1], [0, 1], [0, 1]]
+
+    elif input_shape == (2, ):
+        paths = {'observations': [[0, 0], [0, 1], [1, 0], [1, 1]]}
+        expected = [[1, 0], [0, 1], [0, 1], [1, 0]]
+
+    return paths, expected
+
+
+class TestCategoricalMLPRegressorWithModel(TfGraphTestCase):
+    @params(((1, ), 2), ((2, ), 2))
+    def test_fit_normalized(self, input_shape, output_dim):
```

Comment:
What happen if input shape is (1,1)?

Reply:
Ryan and I discussed about this. Since the underlying model for this regressor is a MLP, a multi-dimensional input does not make much sense and it is assumed that the input is flattened. Currently none of the regressors implicitly flatten the input. This needs to be noted down somewhere that the input is expected to be of shape (x, ).

Reply:
I see. I think adding a checking / assert would be better in this case, for all MLP regressors.

Reply:
That's a good idea.

Reply:
Currently all MLPBaselines assume single dimension for input by passing `env_spec.observation_space.flat_dim` as the `input_shape`. However, they do not flatten the inputs in `fit()` and `predict()`. If a MLPBaseline is used in a multi-dimensional environment, would that be a problem? @ryanjulian

Reply:
i think it is fine. you just need to make it clear in the docs that the interface assumes flattened inputs. in the future i think we should flatten them automatically.

Reply:
Alright, I've updated the docs to include the note on the flattened inputs.